### ISSUE TITLE
addIceCandidate end of candidates woes

### ISSUE THREAD

**Issue by fippo** [w3c/webrtc-pc#1213]
there has been much back-and-forth on this... @taylor-b you might know.
How do I make an end-of-candidates indications for all transports that need one?

http://w3c.github.io/webrtc-pc/#dom-rtcpeerconnection-addicecandidate
> If candidate.candidate is an empty string, process candidate as an end-of-candidates indication for the corresponding media description and ICE candidate generation.

This assumes a end of candidates indication for a particular sdpMid/sdpMLineIndex however.

However, http://w3c.github.io/webrtc-pc/#dom-rtcpeerconnectioniceevent still says that the candidate (and therefore the sdpMId and sdpMLineIndex) is null on end-of-candidates and there is only one.

**Comment by fippo** [comment]
in particular the current spec breaks the assumption that you can hook up pc.addIceCandidate(e.candidate) with pc.onicecandidate.

**Comment by taylor-b** [comment]
There isn't, and this was intentional. The assumption was that people currently use the `null` candidate as an "end of gathering" indication (since Chrome didn't implement `icegatheringstatechange` for a while), and in their `icecandidate` event handler they check `if (candidate)` before signaling it. So by making the "end-of-candidates" indication an empty string, the hope was that it will just start working one day without application developers needing to do anything.

In summary: There are three things that come out of the `icecandidate` event:

* `{sdpMid: "foo", sdpMLineIndex: 0, candidate:"candidate:449267254 1 udp 2122194687 ..."}`
* `{sdpMid: "foo", sdpMLineIndex: 0, candidate:""}`
* `null`

And only the first two can also be passed into `addIceCandidate`.

So, beyond summarizing this somewhere (which is very much needed), is there anything further that needs to be done?

**Comment by fippo** [comment]
If we do bad hacks then I don't see a reason not to allow null in addIceCandidate.

I have written code that assumes that for any candidate event where e.candidate is set the candidate can be parsed. This code will probably throw an exception (inside the event handler) when trying to parse an empty string.

You might also want to revert https://github.com/webrtc/samples/pull/866 if you think it is incorrect.

**Comment by soareschen** [comment]
> There are three things that come out of the icecandidate event

I am confused now. Is there any place in the spec where the second case of empty string candidate must be fired? Chrome and Firefox currently only fire case 1 and 3.

The spec need to clarify the sequence of events expected to be fired. And if null candidate is there for legacy reason and listening to `icegatheringstatechange` is the better approach, there should be a note to recommend so.

Even the example 13 in section 11.2 of the spec do not consider the case for null candidate.

**Comment by aboba** [comment]
@fippo Can you submit a PR?
